sockeye/training.py

```diff
@@ -228,40 +275,141 @@ def fit(self,
                         self.config.max_updates,
                         self.config.max_checkpoints)
 
+        # TODO: support en-fr and fr-en steps
+
+        # TODO: train bi-directional models whenever source/target languages are given!??
+        # Note: for now we do bi-directional updates in the presence of mono data and uni-directional updates without them we should make this more flexible... (e.g. by having mt-steps being specified on the CLI?!)
+        if mono_iters is not None and len(mono_iters) > 1:
+            steps = []
+            num_add_mt_steps = 1
+            assert mass_steps or bt_steps, "Either MASS steps or BT steps must be turned on when monolingual data is supplied"
+            for lang in mono_iters:
+                if mass_steps:
+                    steps.append(("MASS", lang))
+                if bt_steps:
+                    if lang == source_lang:
+                        bt_src_lang = target_lang
+                        bt_trg_lang = source_lang
+                    elif lang == target_lang:
+                        bt_src_lang = source_lang
+                        bt_trg_lang = target_lang
+                    else:
+                        assert False
+                    steps.append(("BT", (lang, bt_src_lang, bt_trg_lang)))
+
+            mt_steps = [
+                ("MT", (source_lang, target_lang)),
+                ("MT", (target_lang, source_lang))
+            ]
+            for _ in range(num_add_mt_steps):
+                steps.extend(mt_steps)
+        else:
+            # TODO: make the use of bidirectional models configurable
+            if source_lang is not None and target_lang is not None:
+                logger.info("Source/target languages given. Enabling bi-directional models")
+                mt_steps = [
+                    ("MT", (source_lang, target_lang)),
+                    ("MT", (target_lang, source_lang))
+                ]
+            else:
+                mt_steps = [("MT", (source_lang, target_lang))]
+            steps = mt_steps
+        loss_prefixes = set()
+
+        random.shuffle(steps)
```

Comment:
so steps contains roughly how many items? Just a few, right? Are we shuffling this list again or is there a good reason to keep the permutation generated once?

Reply:
Yes, only a few items. It is shuffled only once. Would you prefer to shuffle it regularly?

Reply:
I have no good intuition about this, but I would probably shuffle it regularly to avoid any kind of consistent bias in the training algorithm towards grouping/not grouping monolingual and parallel steps if you train for a long time. Not sure, just speculating.